### DISCUSSION TITLE
Searching line-wrapped paragraphs

### DISCUSSION THREAD

**Discussion by jeanas** [BurntSushi/ripgrep#2485] [General] [answered]
Suppose I read some documentation and find a typo or want to improve something *at the exact point I’m looking at*. I can try

```
rg "a few words around that point"
```

to locate the source. However, documentation files (be it md, rst, …) typically consist of paragraphs with line breaks. So in fact I need

```
rg -U 'a\sfew\swords\saround\sthe\spattern'
```

I find myself in this situation pretty often. Would there be a point in adding a “paragraph mode” to ripgrep? In such a mode, single line breaks would essentially be treated as spaces for the purpose of matching. Or did I miss the magic switch for this in the documentation?

**Comment by BurntSushi** [comment, marked as answer]
I also run into this problem occasionally, but it's not just because of paragraphs. It also happens with, for example, string literals. In Rust:

```rust
let foo =
    "This string is \
     split across \
     multiple lines.";
```

And in Go:

```go
foo := "this string is" +
    "split across" +
    "multiple lines"
```

And in this case, the suggested paragraph mode doesn't really work. Indeed, there really isn't a single approach for how string literals might be wrapped. But I run into this sort of case myself too.

To be honest, what I usually do is just change my query to use fewer words under the assumption that there is a break between them somewhere. And that usually solves my problem. Your solution for paragraph breaks specifically is good.

Otherwise, I don't think it's worth adding a special purpose feature for this to ripgrep. It would be difficult to discover and would only work in certain cases.

**Reply by jeanas** [reply]
Well, searching for literal strings in programs is indeed a difficult problem (and [there exist](https://www.kernel.org/doc/html/v4.10/process/coding-style.html#breaking-long-lines-and-strings) coding conventions that recommend against wrapping potentially greppable strings for exactly this reason).


However, considering the huge amount of documentation and other text files written in markup languages that support “1 line break is the same as a space, 2 line breaks end a paragrah” (i.e. every markup language that I know of, like Markdown, reST, Texinfo, LaTeX, etc), I still think it could be worth it. But it’s your call, obviously.

**Reply by jeanas** [reply]
P.S. git diff's “word diff” mode does something similar IIRC (but I didn’t find a way to do this type of search in git grep).
